### PR TITLE
Fixed version conflict logic

### PR DESCRIPTION
**Pull request by davidfowl**
- Added tests
- Show conflicts and the paths to where conflicts occur

https://github.com/NuGet/Home/issues/968

/cc @pranavkm @emgarten @anurse @yishaigalatzer

### CODE REVIEW

File: src/NuGet.Commands/RestoreGraph.cs [hunk 1]
```diff
@@ -101,13 +105,16 @@ public static RestoreTargetGraph Create(bool writeToLockFile, IEnumerable<GraphN
 
             var conflicts = new Dictionary<string, HashSet<ResolverRequest>>();
             var downgrades = new List<Tuple<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>>>();
+            var versionConflicts = new List<Tuple<GraphNode<RemoteResolveResult>, GraphNode<RemoteResolveResult>>>();
```
Comment (pranavkm) [inline src/NuGet.Commands/RestoreGraph.cs]: Code cleanup suggestion: Why not just have these be `VersionConflictResult`, `DowngradeResult` etc? The `Tuple` doesn't do anything useful.

File: src/NuGet.Commands/VersionConflictResult.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Comment (pranavkm) [inline src/NuGet.Commands/VersionConflictResult.cs]: Remove and sort.

File: src/NuGet.DependencyResolver.Core/GraphModel/AnalyzeResult.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace NuGet.DependencyResolver
+{
+    public class AnalyzeResult<TItem>
+    {
+        public List<Tuple<GraphNode<TItem>, GraphNode<TItem>>> Downgrades { get; private set; }
```
Comment (pranavkm) [inline src/NuGet.DependencyResolver.Core/GraphModel/AnalyzeResult.cs]: Here: `List<DowngradeResult> Downgrades { get; }` etc. Looks much cleaner.

Comment (pranavkm) [inline src/NuGet.DependencyResolver.Core/GraphModel/AnalyzeResult.cs]: Nevermind. `Analyze` is generic. One of these days you have to explain why it's so.

File: test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs [hunk 4]
```diff
@@ -15,6 +15,294 @@ namespace NuGet.DependencyResolver.Tests
...
+            provider.Package("B", "2.0")
+                    .DependsOn("A", "1.0");
+
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+            var node = await DoWalkAsync(walker, "A");
+
+            var result = node.Analyze();
+
+            Assert.Equal(1, result.Cycles.Count);
```
Comment (pranavkm) [inline test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs:33]: For future reference. Assert.Collection looks a bit nicer:

```
Assert.Collection(result.Cycles,
   cycle => { AssertPath(cycle, "A 1.0", "B 2.0", "A 1.0"); }
);
```

File: test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs [hunk 5]
```diff
@@ -15,6 +15,294 @@ namespace NuGet.DependencyResolver.Tests
...
+
+            var cycle = result.Cycles[0];
+
+            AssertPath(cycle, "A 1.0", "B 2.0", "A 5.0");
+        }
+
```
Comment (pranavkm) [inline test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs:63]: Extra line.

File: test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs [hunk 6]
```diff
@@ -15,6 +15,294 @@ namespace NuGet.DependencyResolver.Tests
...
+            var result = node.Analyze();
+
+            Assert.Equal(1, result.Cycles.Count);
+
+            var cycle = result.Cycles[0];
+
+            AssertPath(cycle, "A 1.0", "B 2.0", "A 1.0");
+        }
+
+        [Fact]
```
Comment (pranavkm) [inline test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs:40]: Also add one where the graph is a bit more nested `A -> B -> C -> D -> A`?

**Comment by pranavkm** [comment]
:shipit: . Would love for @anurse to also look at this before we check it in.

File: src/NuGet.Commands/RestoreCommand.cs [hunk 7]
```diff
@@ -92,6 +92,12 @@ public Task<RestoreResult> ExecuteAsync()
                     _log.LogError($"Cycle detected {Environment.NewLine} {cycle.GetPath()}");
                 }
 
+                foreach (var versionConflict in g.VersionConflicts)
+                {
+                    _success = false;
+                    _log.LogError($"Version conflict detected for {versionConflict.Selected.Key.Name}. {Environment.NewLine} {versionConflict.Selected.GetPath()} {Environment.NewLine} {versionConflict.Conflicting.GetPath()}");
```
Comment (emgarten) [inline src/NuGet.Commands/RestoreCommand.cs]: Add a resource string for this error

File: test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs (hunk 4 above)
Comment (davidfowl) [inline test/NuGet.DependencyResolver.Tests/RemoteDependencyWalkerTests.cs:33]: That's pretty nice.

File: test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs [hunk 8]
```diff
@@ -1,138 +0,0 @@
...
-
-            // Assert
-            Assert.True(result);
-
-            Assert.Equal(Disposition.Rejected, nodeC138.Disposition);
-            Assert.Equal(Disposition.Accepted, nodeC180.Disposition);
-        }
-
-        [Fact]
-        public void TryResolveConflicts_WorksWhenVersionRangeIsNotSpecified()
```
Comment (emgarten) [inline test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs:57]: This doesn't appear to be covered in the new tests.

File: test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs [hunk 9]
```diff
@@ -1,138 +0,0 @@
...
-            // Assert
-            Assert.True(result);
-
-            Assert.Equal(Disposition.Accepted, nodeCWithEmptyVersionRange.Disposition);
-            Assert.Equal(Disposition.Rejected, nodeC180.Disposition);
-        }
-
-        [Fact]
-        public void TryResolveConflicts_ThrowsErrorsForAllUnresolvableConflicts()
-        {
```
Comment (emgarten) [inline test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs:92]: Do we still display all errors? I don't see a test replacing this.

Comment (davidfowl) [inline test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs:92]: Errors are logs, no longer exceptions. It's the same for cycles, downgrades and version conflicts.

File: test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs (hunk 8 above)
Comment (davidfowl) [inline test/NuGet.DependencyResolver.Core.Tests/GraphOperationsTest.cs:57]: Will add a test

**Comment by emgarten** [comment]
:shipit:

**Comment by davidfowl** [comment]
Ok do one more pass, I think I'm done with this change. I addressed most of the feedback.

**Comment by yishaigalatzer** [comment]
@davidfowl this was merged?

**Comment by emgarten** [comment]
I merged this into 3.1.2, I'll close this PR.

**Pull request closed without merging**